sysl2/sysl/importer/xsd.go

```diff
@@ -0,0 +1,224 @@
+		if name.Local == "_self" {
+			rootType := data.(*xsd.ComplexType)
+			data = rootType.Elements[0].Type
+			t := FindType(data, &types)
+			if t == nil {
+				t = makeType(name, data, &types, logger)
+				if name.Space != "" {
+					types.Add(makeNamespacedType(name, t))
+				}
+			}
+			x := t.(*StandardType)
+			if !contains("~xml_root", x.Attributes) {
+				x.Attributes = append(x.Attributes, "~xml_root")
+			}
+		} else if t := FindType(data, &types); t == nil {
+			types.Add(makeType(name, data, &types, logger))
+		}
+	}
+
+	return types
+}
+
+func FindType(t xsd.Type, knownTypes *TypeList) Type {
+	if res, found := knownTypes.Find(fmt.Sprintf("%s:%s", xsd.XMLName(t).Space, xsd.XMLName(t).Local)); found {
+		return res
+	}
+	if res, found := knownTypes.Find(xsd.XMLName(t).Local); found {
+		return res
+	}
+	return nil
+}
+
+func makeType(_ xml.Name, from xsd.Type, knownTypes *TypeList, logger *logrus.Logger) Type {
+	switch t := from.(type) {
+	case *xsd.ComplexType:
+		return makeComplexType(t, knownTypes, logger)
+	case *xsd.SimpleType:
+		return makeSimpleType(t, knownTypes, logger)
+	case xsd.Builtin:
+		return makeXsdBuiltinType(t, knownTypes)
+	}
+	return nil
+}
+
+func makeComplexType(from *xsd.ComplexType, knownTypes *TypeList, logger *logrus.Logger) Type {
+	createChildItem := func(name xml.Name, data xsd.Type, isAttr, optional, plural bool) Field {
+		childType := FindType(data, knownTypes)
+		if childType == nil {
+			childType = makeType(name, data, knownTypes, logger)
+			knownTypes.Add(childType)
+		}
+		f := Field{
+			Name: name.Local,
+			Type: childType,
+		}
+		if isAttr {
+			f.Attributes = []string{"~xml_attribute"}
+		}
+		f.Optional = optional
+
+		if from, ok := data.(*xsd.SimpleType); ok && from.Restriction.Min > 0 {
+			spec := sizeSpec{
+				Min: int(from.Restriction.Min),
+				Max: int(from.Restriction.Max),
+			}
+			if spec.Max > 0 {
+				spec.MaxType = MaxSpecified
+			}
+			f.SizeSpec = &spec
+		}
+
+		if plural {
+			f.Type = &Array{Items: f.Type}
+		}
+		return f
+	}
+
+	item := &StandardType{
+		name: from.Name.Local,
+	}
+
+	for _, child := range from.Elements {
+		c := createChildItem(child.Name, child.Type, false, child.Optional, child.Plural)
+		if c.SizeSpec == nil {
+			c.SizeSpec = makeSizeSpecFromAttrs(child.Attr)
+		}
+		item.Properties = append(item.Properties, c)
+	}
+	for _, child := range from.Attributes {
+		c := createChildItem(child.Name, child.Type, true, child.Optional, child.Plural)
+		if c.SizeSpec == nil {
+			c.SizeSpec = makeSizeSpecFromAttrs(child.Attr)
+		}
+		item.Properties = append(item.Properties, c)
+	}
+
+	return item
+}
+
+func makeSizeSpecFromAttrs(attrs []xml.Attr) *sizeSpec {
+	ss := sizeSpec{}
+	changed := false
+	for _, attr := range attrs {
+		switch attr.Name.Local {
+		case "maxOccurs":
+			if attr.Value == "unbounded" {
+				ss.MaxType = OpenEnded
+				changed = true
+			} else if val, err := strconv.ParseInt(attr.Value, 0, 32); err == nil {
```

Comment:
What happens when err is not nil?

Reply:
A warning would help that the maxOccurs is not a number in the XSD.

Reply:
I'm going on leave for 2 days and want this merged, opened a issue to deal with it aftr